In [16]:
import pandas as pd

df = pd.read_csv("../test_data/outputs/Left20ms30up_output.tsv", sep="\t", decimal=",", low_memory=False)

gt_col = "gt_event_type" if "gt_event_type" in df.columns else "Eye movement type"
pred_col = "ivt_event_type_smoothed" if "ivt_event_type_smoothed" in df.columns else "ivt_event_type"

valid_labels = {"Fixation", "Saccade"}
mask_mismatch = (
    df[gt_col].isin(valid_labels)
    & df[pred_col].isin(valid_labels)
    & (df[gt_col].astype(str) != df[pred_col].astype(str))
)

mismatch_idx = df.index[mask_mismatch].to_list()
len(mismatch_idx)

context = 1
n_total = len(df)

indices = set()
for i in mismatch_idx:
    start = max(0, i - context)
    end = min(n_total - 1, i + context)
    indices.update(range(start, end + 1))

df_ctx = df.loc[sorted(indices)].copy()

cols = []
if "time_ms" in df_ctx.columns:
    cols.append("time_ms")
if "velocity_deg_per_sec" in df_ctx.columns:
    cols.append("velocity_deg_per_sec")
cols += [gt_col, pred_col]

df_min = df_ctx[cols].copy()
df_min = df_min.rename(columns={gt_col: "gt_class", pred_col: "ivt_class"})

if "time_ms" in df_min.columns:
    df_min = df_min.sort_values("time_ms")

df_min.head(60)


time_ms  velocity_deg_per_sec  gt_class ivt_class
755       2720                 58.33   Saccade   Saccade
756       2724                 31.30  Fixation   Saccade
757       2727                 19.03  Fixation  Fixation
6992     23510                 36.53   Saccade   Saccade
6993     23514                 29.45   Saccade  Fixation
6994     23517                 26.80  Fixation  Fixation
7408     24897                 62.78   Saccade   Saccade
7409     24900                 28.93   Saccade  Fixation
7410     24904                 10.47  Fixation  Fixation
8432     28312                 19.12  Fixation  Fixation
8433     28315                 29.31   Saccade  Fixation
8434     28319                 44.87   Saccade   Saccade
9481     31882                 35.78   Saccade   Saccade
9482     31885                 30.52  Fixation   Saccade
9483     31889                 29.18  Fixation  Fixation
9484     31892                 34.11   Saccade   Saccade
9485     31895                 30.84  Fixation   Saccade
9486     31899                 23.54  Fixation  Fixation
9487     31902                 31.01  Fixation   Saccade
9488     31905                 31.95   Saccade   Saccade
9496     31932                 36.56   Saccade   Saccade
9497     31935                 30.82  Fixation   Saccade
9498     31939                 34.26   Saccade   Saccade
10915    36875                 31.43   Saccade   Saccade
10916    36879                 29.89   Saccade  Fixation
10917    36882                 26.31  Fixation  Fixation
12083    40819                 28.00  Fixation  Fixation
12084    40822                 30.09  Fixation   Saccade
12085    40825                 27.29  Fixation  Fixation
12419    41964                 22.71  Fixation  Fixation
12420    41967                 31.20  Fixation   Saccade
12421    41970                 10.14  Fixation  Fixation
13079    44184                 37.06   Saccade   Saccade
13080    44187                 30.16  Fixation   Saccade
13081    44190                 31.94   Saccade   Saccade
13084    44200                 29.80  Fixation  Fixation
13085    44204                 30.35  Fixation   Saccade
13086    44207                 33.44   Saccade   Saccade
13119    44317                 21.39  Fixation  Fixation
13120    44320                 31.60  Fixation   Saccade
13121    44324                 31.18  Fixation   Saccade
13122    44327                 33.95   Saccade   Saccade
13128    44347                 42.58   Saccade   Saccade
13129    44350                 30.08  Fixation   Saccade
13130    44354                 38.07   Saccade   Saccade
13468    45485                 21.90  Fixation  Fixation
13469    45489                 34.39  Fixation   Saccade
13470    45492                  3.32  Fixation  Fixation
13799    46629                 40.59   Saccade   Saccade
13800    46632                 30.39  Fixation   Saccade
13801    46635                 41.59   Saccade   Saccade
14272    48205                 31.24   Saccade   Saccade
14273    48209                 30.56  Fixation   Saccade
14274    48212                 26.47  Fixation  Fixation
16553    55809                 34.33   Saccade   Saccade
16554    55812                 30.21  Fixation   Saccade
16555    55815                 33.99   Saccade   Saccade
17547    59122                 10.63  Fixation  Fixation
17548    59125                 30.14  Fixation   Saccade
17549    59129                 66.08   Saccade   Saccade

# Analysis: 300Hz Dataset Errors

## Key Findings

**Total Errors:** 63/68,582 samples (99.91% accuracy)
- Fixation→Saccade: 35 errors
- Saccade→Fixation: 28 errors

### Pattern 1: Event Boundary Clustering
- **81% of errors** (51/63) occur at event boundaries
- GT transitions: S→S→F (15), F→F→S (12), F→S→S (11), S→F→F (9)
- Most are **last saccade sample** or **first fixation sample** at transitions

### Pattern 2: Near-Threshold Distribution
- Median error velocity: **30.32°/s** (threshold: 30.2°/s)
- 22 errors within 0.5°/s of threshold
- 31 errors within 1.0°/s of threshold
- 39 errors within 2.0°/s of threshold

### Pattern 3: Acceleration Signature
- **Fixation→Saccade errors** (wrongly kept as Fixation):
  - Mean acceleration: **+3,268 deg/s²** (speeding up)
  - 71% have positive acceleration
  - Median: +5,087 deg/s²
  
- **Saccade→Fixation errors** (wrongly kept as Saccade):
  - Mean acceleration: **-202 deg/s²** (slowing down)
  - 54% have negative acceleration
  - Median: -567 deg/s²

### Pattern 4: GT vs Velocity Mismatch
Many errors show **ground truth labels contradicting velocity**:
- GT="Saccade" but v=28.9°/s (below threshold)
- GT="Fixation" but v=31.0°/s (above threshold)
- GT="Saccade" but v=29.9°/s (below threshold)

This suggests **inherent ambiguity** in transition samples where:
1. Human annotators use temporal context (event continuity)
2. Velocity-based classifier uses instantaneous measurement
3. Both are "correct" given their information

## Tested Strategies (All Increased Errors)

1. **Dynamic margin** (dist-based confidence): 62→63 errors ❌
2. **Acceleration rule** (accel>3000 → Saccade): 62→167 errors ❌
3. **Acceleration+context** (accel + neighbors): 62→105 errors ❌
4. **Run-length smoothing** (isolated outlier removal): 62→95 errors ❌

## Conclusion

The remaining 62 errors represent the **accuracy limit** of threshold-based IVT classification:
- Ground truth uses event-level context (transitions span multiple samples)
- IVT uses sample-level velocity (instantaneous, no temporal context)
- Boundary samples are **fundamentally ambiguous** without higher-level features

Further improvement would require:
- Machine learning classifier with temporal features
- Event-level post-processing (HMM, CRF)
- Multi-modal features (acceleration, pupil, eye model confidence)